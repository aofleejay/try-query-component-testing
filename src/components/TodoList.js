import React, { Fragment } from 'react'
import { Query, Mutation } from "react-apollo"
import gql from "graphql-tag"

export const GET_ALL_TODOES = gql`
  query allTodoes {
    allTodoes {
      id
      content
    }
  }
`

export const CREATE_TODO = gql`
  mutation createTodo($content: String!, $isDone: Boolean!) {
    createTodo(content: $content, isDone: $isDone) {
      id
      content
    }
  }
`

const TodoList = (props) => (
  <Fragment>
    <Query query={GET_ALL_TODOES}>
      {({ loading, error, data }) => {
        if (loading) return <p>Loading...</p>
        if (error) return <p>Error :(</p>

        return data.allTodoes.map(({ id, content }) => (
          <div key={id}>
            <p>{content}</p>
          </div>
        ))
      }}
    </Query>
    <Mutation mutation={CREATE_TODO}>
      {(createTodo, { data }) => (
        <form
          onSubmit={(e) => {
            e.preventDefault()
            createTodo({
              variables: { content: e.target.text.value, isDone: false },
              refetchQueries: [{ query: GET_ALL_TODOES }],
            })
          }}
        >
          <input id="text" type="text" name="text" />
          <button id="submit" type="submit">Add Todo</button>
        </form>
      )}
    </Mutation>
  </Fragment>
)

export default TodoList
