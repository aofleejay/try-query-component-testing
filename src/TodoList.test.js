import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import renderer from 'react-test-renderer'
import waait from 'waait'
import TodoList, { GET_ALL_TODOES } from './TodoList'

describe('Test TodoList component', () => {
  it('Should contain loading ui when promise is pending', () => {
    const component = renderer.create(
      <MockedProvider mocks={[]} addTypename={false}>
        <TodoList />
      </MockedProvider>,
    )
  
    expect(component).toMatchSnapshot()
  })

  it('Should contain response when promise has resolved', async () => {
    const mocks = [
      {
        request: {
          query: GET_ALL_TODOES,
        },
        result: {
          data: {
            allTodoes: [
              {
                id: 'cjiugybsw23z8019717p0cgwb',
                content: 'Write a blog',
              },
              {
                id: 'cjiugyxca23zc0197lxs4uh29',
                content: 'Practice guitar',
              },
            ],
          },
        },
      },
    ]

    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TodoList />
      </MockedProvider>,
    )

    await waait(0)
  
    expect(component).toMatchSnapshot()
  })

  it('Should contain error when promise has rejects', async () => {
    const mocks = [
      {
        request: {
          query: GET_ALL_TODOES,
        },
        error: new Error('Noooooooo!!!!')
      },
    ]

    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TodoList />
      </MockedProvider>,
    )

    await waait(0)
  
    expect(component).toMatchSnapshot()
  })
})
