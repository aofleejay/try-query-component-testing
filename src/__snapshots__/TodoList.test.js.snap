// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoList component Should contain error when promise has rejects 1`] = `
Array [
  <p>
    Error :(
</p>,
  <form
    onSubmit={[Function]}
>
    <input
        name="text"
    />
    <button
        type="submit"
    >
        Add Todo
    </button>
</form>,
]
`;

exports[`Test TodoList component Should contain loading ui when promise is pending 1`] = `
Array [
  <p>
    Loading...
</p>,
  <form
    onSubmit={[Function]}
>
    <input
        name="text"
    />
    <button
        type="submit"
    >
        Add Todo
    </button>
</form>,
]
`;

exports[`Test TodoList component Should contain response when promise has resolved 1`] = `
Array [
  <div>
    <p>
        Write a blog
    </p>
</div>,
  <div>
    <p>
        Practice guitar
    </p>
</div>,
  <form
    onSubmit={[Function]}
>
    <input
        name="text"
    />
    <button
        type="submit"
    >
        Add Todo
    </button>
</form>,
]
`;
